File: 2022/day05_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from typing import Any

type Instruction = tuple[int, int, int]
type Stacks = list[list[str]]


def solution(content: str) -> Any:
    stacks, instructions = parse(content)
    for instruction in instructions:
        stacks = act(instruction, stacks)
    return tops(stacks)


def tops(stacks: Stacks):
    return "".join(stack[0] for stack in stacks)


def act(instruction: Instruction, stacks: Stacks) -> Stacks:
    qty, frm, to = instruction
    copy = stacks.copy()
    while qty > 0:
        fst, *rest = copy[frm]
        copy[frm] = rest
        copy[to].insert(0, fst)
        qty -= 1
    return copy


def parse(content: str):
    part_1, part_2 = content.split("\n\n")
    return (parse_stacks(part_1), parse_instructions(part_2))


def parse_stacks(content: str) -> Stacks:
    lines = content.split("\n")
    size = (len(lines[0]) + 1) // 4
    stacks: list[str] = ["" for _ in range(size)]
    for line in lines[:-1]:
        for idx in range(size):
            index = idx * 4 + 1
            crate = line[index].strip()
            if crate:
                stacks[idx] += crate
    return [list(stack) for stack in stacks]


def parse_instructions(content: str):
    return [parse_procedure(line) for line in content.split("\n") if line]


def parse_procedure(line: str) -> Instruction:
    match = re.match(r"move (\d+) from (\d+) to (\d+)", line)
    if match is None:
        raise ValueError(f"Invalid line {line}")
    return (int(match[1]), int(match[2]) - 1, int(match[3]) - 1)


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.read()))


if __name__ == "__main__":
    main()

File: 2022/day05-1.ts
type Instruction = {
  qty: number;
  from: number;
  to: number;
};

type Stacks = string[][];

function solution(content: string) {
  const [stacks, instructions] = parse(content);

  return tops(instructions.reduce((acc, cur) => act(cur, acc), stacks));
}

function tops(stacks: Stacks) {
  return stacks.map((stack) => stack[0]).join("");
}

function act(instruction: Instruction, stacks: Stacks): Stacks {
  const { qty, from, to } = instruction;
  const copy = stacks.slice();

  const toMove = copy[from].splice(0, qty).toReversed();
  copy[to] = [...toMove, ...copy[to]];

  return copy;
}

function parse(content: string): [Stacks, Instruction[]] {
  const [part1, part2] = content.split("\n\n");
  return [parseStacks(part1), parseInstructions(part2)];
}

function parseStacks(content: string): Stacks {
  const lines = content.split("\n");
  const size = (lines[0].length + 1) / 4;
  const stacks: Stacks = Array.from({ length: size }).map(() => []);
  lines.slice(0, -1).forEach((line) => {
    Array.from({ length: size })
      .map((_, idx) => idx * 4 + 1)
      .forEach((index, idx) => {
        const crate = line[index].trim();
        if (crate) {
          stacks[idx].push(crate);
        }
      });
  });
  return stacks;
}

function parseInstructions(content: string) {
  return content
    .split("\n")
    .filter((line) => line)
    .map(parseProcedure);
}

function parseProcedure(line: string): Instruction {
  const match = line.match(/move (\d+) from (\d+) to (\d+)/);
  if (match === null) throw new Error(`Invalid line: ${line}`);
  return {
    qty: Number(match[1]),
    from: Number(match[2]) - 1,
    to: Number(match[3]) - 1,
  };
}

Deno.readTextFile("day05.dat")
  .then((content) => content)
  .then(solution)
  .then(console.log);
